File: backend/app/schemas/tracking.py
from datetime import date, datetime
from typing import Optional
from uuid import UUID

# pyrefly: ignore [missing-import]
from pydantic import BaseModel, ConfigDict


# ==========================================================
# Tracking Timeline Event
# ==========================================================

class TrackingTimelineEvent(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    status: str
    hub: Optional[str] = None
    remarks: Optional[str] = None
    event_time: datetime


# ==========================================================
# Public Tracking Response
# ==========================================================

class TrackingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tracking_number: str
    status: str
    current_hub: Optional[str] = None
    expected_delivery_date: Optional[date] = None
    updated_at: datetime
    timeline: list[TrackingTimelineEvent]

File: frontend/src/lib/types/tracking.ts
import type { ShipmentStatus } from "@/lib/types/shipment";

export interface TrackingTimelineEntry {
    id: string;
    status: ShipmentStatus;
    remarks?: string;
    created_at: string;
}

export interface TrackingResult {
    tracking_number: string;
    status: ShipmentStatus;
    current_hub?: string;
    updated_at: string;
    timeline: TrackingTimelineEntry[];
}

export interface TrackingTimelineEvent {
    id: string;
    status: ShipmentStatus;
    hub?: string;
    remarks?: string;
    event_time: string;
}

export interface TrackingDetail {
    tracking_number: string;
    status: ShipmentStatus;
    current_hub?: string;
    expected_delivery_date?: string;
    updated_at: string;
    timeline: TrackingTimelineEvent[];
}